Changed condition-style and added more stuff

TornaxO7 opened the pull request:
Before you about merging this, I'd like to discuss first about this pull request since it changes a lot which means that many things might break as well! But before that, here's a short summary what I've changed:

- added `doc/cosco.vim` for `:h cosco`
- changed the conditions of `cosco`. It's more "inspired" from the C language.
    I changed the conditions so the auto-setting works "better" since the autosetting was a little bit problematic for me as a C dev:
    ![GIF](https://user-images.githubusercontent.com/50843046/106335541-528fd200-628d-11eb-98fb-091aa1c611ea.gif)

    This is now after my changes:
    ![GIF](https://user-images.githubusercontent.com/50843046/106336467-40af2e80-628f-11eb-9153-136f66521440.gif)
   
    It's still a little bit buggy in some cases, but it works in the most cases in C.
- the autosetting is also "compatible" with snippet extensions (cosco shouldn't interfere the tabstops)
- autosetting can detect comments
- changed the `CONTRIBUTING.md` file

Now there's another problem:
Your unittests "don't work" since your autosetting approach differs my. So it fails in some cases:
```

Started at Sat 30 Jan 2021 12:11:14 AM CET

------------------------------------------------------------------------------
Cosco.vim

test_nextStartsWithRightCurlyBrace => .
test_prevIsComma_nextIsClosingBlock => .F
test_prevIsComma_nextIsComma => FFF
test_prevIsComma_nextIsLessIndented => FFF
test_prevIsComma_nextIsSameIndentation => FFF
test_prevIsLeftBracket_nextIsWhatever => *
test_prevIsLeftBracket_nextStartsWithRightBracket => *
test_prevIsLeftCurlyBrace_nextIsComma => *
test_prevIsLeftCurlyBrace_nextIsVarComma => *
test_prevIsLeftCurlyBrace_nextIsWhatever => *
test_prevIsLeftParens_nextIsWhatever => *
test_prevIsLeftParens_nextStartsWithRightParens => *
test_prevIsVarLeftCurlyBrace_nextIsRightCurlyBrace => *
test_prevSC_nextSC => FFF
```
So we'd need to discuss, how we'd have to change them or the conditions. I let my friends test it a little bit with rust and C++. The autosetting works mostly for them.

If you want to know more what has changed, than you can take a look into the CONTRIBUTING.md file. I added some information about it and of course, you can read through each file :)

TornaxO7 committed:
Huge changes!

TornaxO7 committed:
Bug fix: Snippet jump should be possible now

TornaxO7 committed:
Feature added:
Whitelist is gonna be read now in order to set the enabled filetypes where the plugin should be active

TornaxO7 committed:
Changed the adding 'style

' the commas/semicolons are now added *after* hitting the return key!

TornaxO7 committed:
Added auto-setting command (:CoscoToggleAutoSetter)

TornaxO7 committed:
fixed misspelled variable: g:cosco_whitelist

TornaxO7 committed:
Fixed bug with rustFuncCall (when writing code in documentstring-markdown parts)

TornaxO7 committed:
Big changes: Improved code, style and cosco workflow

TornaxO7 committed:
Added condition, that no semicolon is added after an curvey bracket

TornaxO7 committed:
multiline conditions won't get semicolon/comma as well

TornaxO7 committed:
Fixed bug, where 'test();' was seen as a multiline condition

TornaxO7 committed:
Changed CONTRIBUTING.md and added a documentation file for :help cosco.txt

TornaxO7 committed:
Bugfix: Arguments in function braces again, have a comma

TornaxO7 committed:
Bugfix for C:
when declaring function like that:
    int
    funcname()
don't add a semicolon after 'int'

lfilho commented:
Wow, @TornaxO7 thanks a lot, this is massive with lots of great stuff! Thanks for taking the time!!

I will need a couple of days to review this.

But before anything: this plugin is quite used by others (100 unique clones in the last 2 weeks alone) so I don't want to introduce a breaking change for them or myself without a good reason, but right now it seems (again, I need time to understand everything you've done, so I might be wrong here, forgive me if so) that the reason is to make it work for your personal preferences / for your particular coding flow. Would you say that's correct?

Also, it would be nice to maintain the code style. There's a lot of diff that's merely personal style things and formatting (like `cosco#commaOrSemiColon()` -. `cosco#CommaOrSemiColon()`, the headers on top of each file (more manual work to maintain - i think Git takes care of that for us nowadays, no need for author list/history in the file itself...). So it would be nice if the changes are kept functional only in order to speed up review and maintain consistency. Thoughts? 

I will start reviewing after you answer the above

TornaxO7 commented:
>  Would you say that's correct?

Yes, I agree.

>  i think Git takes care of that for us nowadays, no need for author list/history in the file itself...)

I can agree _mostly_ to that. We can remove the Author and Version part of the headers and the filename as well. But I'd like to keep the Usage part, since it summarize the content of the file if it's fine for you.

TornaxO7 committed:
Changes in autocmd
Put the condition for g:cosco_auto_comma_or_semicolon_events over the autocmd BufEnter instead in the Autocmd-Function for better performance when disabled

TornaxO7 commented:
@lfilho Can you send me a message, when you're starting to review my pull request?

TornaxO7 committed:
Bugfix:
cosco added semicolon after case instruction.
New function: cosco_setter#MakeDoublePoints()
   - adds a ':' at the end of the previous line
   - currently only active if the previous line has the keyword 'case' as the first word in the line

TornaxO7 committed:
Bugfix:
Double points (':') are added now as well, if there's the 'default' keyword in a switch case.

TornaxO7 committed:
Bugfix:
For the following case:
    func1(
       arg1,
       func2(),
       |
    )  ^
     Cursor

Cosco wouldn't go to the last elseif condition, since it only looks, if there's only a semicolon after a round brace.
In this case we've a comma so a short fix of the regex from ');$' to ')[,;]$' should fix this issue.

To sum it up: Cosco shouldn't add a semicolon after 'func2(),' (result would be 'func2(),;') now.

TornaxO7 committed:
Updated the CONTRIBUTION.md

TornaxO7 committed:
DANGER: TEST COMMIT

TornaxO7 committed:
Bugfix:
Following example:
     );
 []

Cosco would add a comma after the previous line due to the square brackets. Should be fixed now
with the new regex pattern.

Also moved the condition if there's already a semicolon/comma to the bottom. But still need to think about the
condition steps so it's only temporary the last condition.

TornaxO7 committed:
- New approach for filterting through conditions: Splitted it up to 5 functions which decide what to do
- Explained it in the CONTRIBUTING.md file
- Adapted the changes (variable names, function names) for each file

TornaxO7 committed:
Fixed variable and function names in the Manual function

TornaxO7 committed:
Fixed wrong substitution

TornaxO7 committed:
Fixed typo in the doc

TornaxO7 committed:
Bugfix:
when typing something like that:

int main() {
    test();
    for (int i=0; i<10; i++) {
        (...)
    }
}

cosco thinks that the open curly bracket of the foor loop is meant to be the curly bracket for a function implementation like that:

    void test()
    {

This should be fixed now by changing the control statement in the ShouldRemove() function.

TornaxO7 committed:
Bugfix:

when typing something like that:

    switch(var) {
        case 1:
            break;

cosco doesn't check, if there's already a ':' so it keeps adding them! This should be fixed now with the new regex pattern.

TornaxO7 committed:
Bugfix:

When user does something like that:

    int main()
    {
        |
    }   ^
     Cursor

=>

    int main()
    { | <-- Cursor

   }

Cosco would think that the 'main()' would be a function call. So we have to add a conditions which looks
if the user is already in a function implementation. This is added in the 'curly bracket' section.

TornaxO7 committed:
Bugfix: Changed the comparison for multiline conditionals. 'stridx' was wrong

TornaxO7 committed:
Replaced matchstr() function with =~ for better performance. See: https://vi.stackexchange.com/questions/29015/fastest-way-to-compare-string-with-pattern for more information

TornaxO7 committed:
Bugfix: Removed uneccessary ending round bracket

TornaxO7 committed:
Bugfix and important change:
Cosco looks now even after the comma, that the indentation is the same as in the previous line.

Rust bugfix: Cosco used to add a semicolon for each line in a struct of rust. This should be fixed now

TornaxO7 committed:
Added description for the rust-struct-condition

TornaxO7 committed:
Commented out the curly-bracket-set-condition because it creates a bug for the following code:

static void
print_hello (GtkWidget* widget,
             gpointer data)
{
    g_print(Sttttttttttttttn), // Here
}

somehow it detect it as a curly-bracket-set. Looks like that this condition isn't needed so I commented it out (temporary).

TornaxO7 committed:
Bugfix:

when entering a function name which includes 'case' or 'default', than cosco thinks, that it's a case or
default instruction in a switch-part.
Should be fixed now

TornaxO7 committed:
Bugfixes:
1. Case:

When the user creates a struct like this:

    struct Structure
    {

then Cosco would add a comma/semicolon after 'Structure'. There's now a condition in the
cosco_eval#ShouldSkip() function, but it'll be probably ported to the 'cosco_eval#Specials()' function.

2. Case:

When the user creates a function like that:

    Test *
    func_name(args)
    {

then Cosco would add a comma/semicolon after the '*', if the user types something in the next line of the star.
This should be fixed now by the new condition in the 'cosco_eval#Specials()' function.

TornaxO7 committed:
Bugfix:
FOr the
For the following case:

    return (
        a
        b, (forgot the comma above)
        c)| <-- Cursor

Cosco would add a comma after the closed brace since it just looks, if 'b' has a comma. This should be fixed now
by the additional condition that the current line doesn't end with a bracket.

TornaxO7 committed:
Fixed some typos in the CONTRIBUTING.md file

TornaxO7 committed:
Bugfix: Cosco used the wrong variable name for skipping comments

TornaxO7 committed:
Bugfix (C++): Following example:

    template <class... T>
    |
   Cursor
Cosco would add a comma after the '>' symbol.
Should be fixed now with the new condition in the c++/c section of the cosco_eval#Specials() function.

TornaxO7 committed:
Outcommented test-line
